### PR TITLE
Fix `Gson.getDelegateAdapter` not working properly for `JsonAdapter`

### LINKED ISSUE #1028: getDelegateAdapter() does not work properly in TypeAdapterFactory used from JsonAdapter annotation
GSON's getDelegateAdapter method does not appear to work properly inside of a TypeAdapterFactory that is used through the \@JsonAdapter annotation.  This was found using GSON 2.8.0 and Java 7.

For example, given a class like:
```Java
public final class Thing {
    @JsonAdapter(NonEmptyMapAdapterFactory.class)
    private final Map<Integer, String> data;

    public Thing() {
        data = new HashMap<>();
    }
}
```

And a TypeAdapterFactory like:
```Java
public class NonEmptyMapAdapterFactory implements TypeAdapterFactory {

    @Override
    public <T> TypeAdapter<T> create(final Gson gson, final TypeToken<T> type) {
        if (!Map.class.isAssignableFrom(type.getRawType())) {
            return null;
        }

        //final TypeAdapter<T> delegate = gson.getAdapter(type);
        final TypeAdapter<T> delegate = gson.getDelegateAdapter(this, type);

        return new TypeAdapter<T>() {
            @Override
            public void write(final JsonWriter writer, T value) throws IOException {
                final Map map = (Map) value;
                if (map == null || map.isEmpty()) {
                    writer.nullValue();
                } else {
                    delegate.write(writer, value);
                }
            }

            @SuppressWarnings("unchecked")
            @Override
            public T read(final JsonReader reader) throws IOException {
                final T map = delegate.read(reader);
                if (map == null) {
                    return (T) new HashMap();
                } else {
                    return map;
                }
            }
        };
    }

}
```

The call to `gson.getDelegateAdapter(this, type);` returns an adapter of type "ReflectiveTypeAdapterFactory$Adapter".  Calling `gson.getAdapter(type);` instead does return the expected "MapTypeAdapterFactory$Adapter", but this obviously won't work correctly if this adapter factory is instead used without the JsonAdapter annotation.  

If the annotation is removed from the field in the Thing class and gson is instead set up like:
```Java
new GsonBuilder()
        .registerTypeAdapterFactory(new NonEmptyMapAdapterFactory())
        .create();
```
Then the custom TypeAdapterFactory shown above works exactly as expected: `gson.getDelegateAdapter(this, type);` returns a "MapTypeAdapterFactory$Adapter", but calling `gson.getAdapter(type);` returns the the same class, resulting in a stack overflow in the read and write methods.

The problem seems to happen because of this code at the beginning of getDelegateAdapter():
```Java
// Hack. If the skipPast factory isn't registered, assume the factory is being requested via
// our @JsonAdapter annotation.
if (!factories.contains(skipPast)) {
  skipPast = jsonAdapterFactory;
}
```
In the list of factories, jsonAdapterFactory comes after the entry for the MapTypeAdapterFactory, causing both to be skipped.

### PR DESCRIPTION
**Pull request by Marcono1234**
### Purpose
Fixes #1028
Supersedes #1690

### Description
Unlike #1690 this implementation here tracks which `TypeAdapterFactory` instances have been created for `@JsonAdapter` placed on classes, and only for them lets `Gson.getDelegateAdapter` skip past the `JsonAdapterAnnotationTypeAdapterFactory`.

This pull request also in general changes the behavior of `Gson.getDelegateAdapter` to hopefully be more reasonable when `skipPast` is `null` (which was previously supported, but undocumented) or when `skipPast` is not a known factory. In these cases the method now behaves like `Gson.getAdapter` (instead of assuming that the factory came from `@JsonAdapter` on a class).

### Checklist

- [x] New code follows the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)
- [ ] If necessary, new public API validates arguments, for example rejects `null`
- [ ] New public API has Javadoc
    - [ ] Javadoc uses `@since $next-version$`  
      (`$next-version$` is a special placeholder which is automatically replaced during release)
- [x] If necessary, new unit tests have been added  
  - [x] Assertions in unit tests use [Truth](https://truth.dev/), see existing tests
  - [x] No JUnit 3 features are used (such as extending class `TestCase`)
  - [x] If this pull request fixes a bug, a new test was added for a situation which failed previously and is now fixed
- [x] `mvn clean verify javadoc:jar` passes without errors

### CODE REVIEW

File: gson/src/main/java/com/google/gson/Gson.java [hunk 1]
```diff
@@ -593,54 +594,71 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
...
+   * }</pre>
+   * Note that this call will skip all factories registered before {@code skipPast}. In case of
+   * multiple TypeAdapterFactories registered it is up to the caller of this function to insure
+   * that the order of registration does not prevent this method from reaching a factory they
+   * would expect to reply from this call.
+   * Note that since you can not override type adapter factories for String and Java primitive
+   * types, our stats factory will not count the number of String or primitives that will be
+   * read or written.
+   *
+   * <p>If {@code skipPast} is {@code null} or a factory which has neither been registered
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/Gson.java:634]: The section starting here is new; the changes above only remove one redundant trailing whitespace (unfortunately the GitHub diff does not indicate that very well).

File: gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java [hunk 2]
```diff
@@ -88,6 +88,19 @@ static String checkInstantiable(Class<?> c) {
...
+    T instance = instanceCreator.createInstance(type);
+    if (instance == null) {
+      throw new RuntimeException("InstanceCreator " + instanceCreator + " returned null for type " + type);
+    }
+
+    if (!rawType.isInstance(instance)) {
+      throw new ClassCastException("InstanceCreator " + instanceCreator + " created instance of wrong type;"
+          + " expected " + rawType.getName() + " but got instance of unrelated type " + instance.getClass().getName());
+    }
+    return instance;
```
Comment (Marcono1234) [inline gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java:101]: Added this for sanity, and to avoid some checks in `JsonAdapterAnnotationTypeAdapterFactory.isClassJsonAdapterFactory` to handle bad `InstanceCreator`s. Most likely Gson was already failing for bad `InstanceCreator`s before, but probably at a completely unrelated place which would have made debugging more difficult.

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
This looks great! Only very minor comments. I'm going to run this through all of Google's internal tests.

**Comment by Marcono1234** [comment]
Thanks! The fact that `JsonAdapterAnnotationTypeAdapterFactory.isClassJsonAdapterFactory` also performs the annotation lookup does not feel that clean to me, but I hope it is fine with the current implementation anyway.

I am a bit afraid that users unwittingly rely on the current broken behavior, or that users are currently intentionally calling `getDelegateAdapter` with an unknown (or `null`) factory to be guaranteed to get the Gson reflection-based adapter (even though this was an implementation detail).
But hopefully your test run shows if that is really a valid concern.

**Review by eamonnmcmanus — APPROVED** [review]

**Comment by eamonnmcmanus** [comment]
Still waiting for test reruns, but looking good so far. (There are so many tests that invariably a few of them fail for unrelated reasons.)

**Comment by Marcono1234** [comment]
Did a [code search on GitHub now](https://github.com/search?q=%22getDelegateAdapter%28null%22&type=code) and unfortunately found a few cases of `getDelegateAdapter(null, ...)`:
- [vertigo-io/vertigo-extensions](https://github.com/vertigo-io/vertigo-extensions/blob/b8788374530c3e44a49ebcd9dc6bea91d2d2d8f0/vertigo-vega/src/main/java/io/vertigo/vega/engines/webservice/json/GoogleJsonEngine.java#L325), [eutro/Tetraits](https://github.com/eutro/Tetraits/blob/22eb0caef86b61436c99391b529e66fa3da46d31/src/main/java/eutros/tetraits/data/gen/SchemaDataProvider.java#L29)
Both seem to work around the limitation that `JsonSerializer` and `JsonDeserializer` do not support delegation, by making the eventually created `Gson` instance accessible through a field and then in the serializer / deserializer use `getDelegateAdapter` with `null` because they don't have access to a `TypeAdapterFactory` there.
I am afraid with the changes of this pull request this would lead to an infinite recursion because the `JsonSerializer` / `JsonDeserializer` would keep retrieving itself.
- [JuiceLabs/kotlin-fhir-model](https://github.com/JuiceLabs/kotlin-fhir-model/blob/ab51504ca9f1189a996764aca56941aa6d8c3270/src/main/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactory.java#L239)
Seems to be a custom variant of gson-extras' `RuntimeTypeAdapterFactory`; not sure why they are using `null` as `skipPast`.

These libraries don't seem to be that popular so maybe the risk of causing issues for them is not that big (we can also warn the authors about their problematic `getDelegateAdapter` usage).

Also, it appears commit 83e5a4937cf13ab9b94a7bb12ce2fefa168752ed which introduced the `getDelegateAdapter` method for Gson did not mention `null` as supported value, and the code from back then looks like it would have thrown an `IllegalArgumentException` for `null` because the workaround for `@JsonAdapter` did not exist back then. Most likely only since b2c00a3b02eec48ed621fa7cfec54ac6da4aa48d a `null` argument would have worked the way it does now. So it seems they are (wittingly?) relying on undocumented Gson implementation details.

**Comment by eamonnmcmanus** [comment]
I'm hesitating a bit here because the internal tests did show up one test failure (or actually a group of failures in related tests, probably having the same cause.

I'm getting a failure like this:
```
java.lang.ClassCastException: InstanceCreator [...].GsonUtils$ImmutableListTypeAdapterFactory$$Lambda$148/0x00000008012f8000@2926ed77 created instance of wrong type; expected com.google.common.collect.ImmutableList but got instance of unrelated type java.util.ArrayList
        at com.google.gson.internal.ConstructorConstructor.useInstanceCreator(ConstructorConstructor.java:99)
        at com.google.gson.internal.ConstructorConstructor$2.construct(ConstructorConstructor.java:127)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:80)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
        at [...].GsonUtils$ImmutableListTypeAdapterFactory$1.read(GsonUtils.java:51)
        at [...].AutoValue_ButtonGroupModel$GsonTypeAdapter.read(AutoValue_ButtonGroupModel.java:84)
        at [...].AutoValue_ButtonGroupModel$GsonTypeAdapter.read(AutoValue_ButtonGroupModel.java:27)
        at com.google.gson.TypeAdapter.fromJson(TypeAdapter.java:269)
        at com.google.gson.TypeAdapter.fromJson(TypeAdapter.java:285)
```

Maybe there is some sort of interaction between this `ImmutableListTypeAdapterFactory` and the `TypeAdapter` created by auto-value-gson. This is `ImmutableListTypeAdapterFactory`:
```java
  private static class ImmutableListTypeAdapterFactory implements TypeAdapterFactory {

    @Nullable
    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
      if (!ImmutableList.class.isAssignableFrom(type.getRawType())) {
        return null;
      }
      final TypeAdapter<T> delegate = gson.getDelegateAdapter(this, type);
      return new TypeAdapter<T>() {
        @Override
        public void write(JsonWriter out, T value) throws IOException {
          delegate.write(out, value);
        }

        @Override
        @SuppressWarnings("unchecked")
        public T read(JsonReader in) throws IOException {
          return (T) ImmutableList.copyOf((List) delegate.read(in));
        }
      };
    }

    public static <E> InstanceCreator<List<E>> newCreator() {
      return type -> new ArrayList<>();
    }
  }
```

Registered like this:
```java
  private static final GsonBuilder GSON_BUILDER = new GsonBuilder();

  static {
    GSON_BUILDER.registerTypeAdapterFactory(new ImmutableListTypeAdapterFactory());
    GSON_BUILDER.registerTypeAdapter(
        ImmutableList.class, ImmutableListTypeAdapterFactory.newCreator());
  }

  public static Gson get() {
    return GSON_BUILDER.create();
  }
```

I'm not sure if this code is doing something unusual. I do notice that the code in `ConstructorConstructor` that is throwing the exception is new with this PR.

That's as much as I had time for here but I may be able to look into it further later. Let me know if this much is enough to get an idea of the problem, and in particular whether it is something other users might run into.

**Review by eamonnmcmanus — CHANGES_REQUESTED** [review]
(Marking this as "request changes" so we don't accidentally merge it before the test-failure issue is resolved.)

**Comment by eamonnmcmanus** [comment]
> Technically the changes to `InstanceCreator` validation are not needed for this pull request, so I can comment them out, add a TODO comment and add `@Ignore` to the corresponding test if you want.

That sounds like a great idea. If those changes are not strictly necessary, I think everything else is pretty uncontroversial. Even taking into account your code-search findings.

We could then have a separate PR for the `InstanceCreator` validation and discuss there whether it is safe.

**Comment by Marcono1234** [comment]
Have reverted the `InstanceCreator` validation changes now.

Though while thinking a bit more about these changes, what do you think about rejecting `null` as `skipPast` for `Gson.getDelegateAdapter`? The handling and support of `null` is undocumented and will be changed by this pull request anyways. Maybe the cleaner solution would then be to reject `null`. I think this would be consistent with most other Gson methods which also don't support `null`, and it might also led to a more helpful exception message than a potential `StackOverflowError` due to infinite recursion for users which currently use `getDelegateAdapter(null, ...)`.

**Comment by Marcono1234** [comment]
@eamonnmcmanus, would it be ok if I changed `getDelegateAdapter` to reject `null` (as proposed above), and could you then please run the Google tests with that change?

**Comment by eamonnmcmanus** [comment]
> @eamonnmcmanus, would it be ok if I changed `getDelegateAdapter` to reject `null` as `skipPast` (as proposed above), and could you then please run the Google tests with that change?

I can certainly try that. Just adding a second `Objects.requireNonNull` at the start of `Gson.getDelegateAdapter`. I see that there are three Gson tests that fail if I do that, but I'll ignore those and see what else might fail.

**Comment by Marcono1234** [comment]
Ah sorry, I should have been more explicit: Rejecting `null` also requires a small change in the `JsonAdapterAnnotationTypeAdapterFactory` class. I have adjusted that now.

If the test results indicate that rejecting `null` is a problem I can revert the latest commit again. (Though I would be a bit curious then, if with `null` being allowed but the behavior change for `null` compared to before, the caller really behaves as expected.)

**Comment by eamonnmcmanus** [comment]
I have confirmed that rejecting `skipPast == null` does not cause any test failures, so I think we can go ahead with it. I can merge it if you confirm it is ready.

**Comment by Marcono1234** [comment]
Thanks a lot for checking! I have made a few small additional adjustments; sorry for the trouble. It should be functionally nearly identical to the previous implementation (except for corner cases with factories with custom `equals`); but if you want you can run the tests again to be sure.

I think these changes should be ready for merging now.

**Review by eamonnmcmanus — APPROVED** [review]
